### ISSUE TITLE
[gltf2] gltf2 export faulting when multiple entities refer to the same mesh

### ISSUE THREAD

**Issue by Misterblue** [assimp/assimp#1653]
This month I've had problems with glTF2Export faulting on some of my scenes. I've traced the problem to what seems to be memory corruption in glTF2Export::MergeMeshes().

On 20171107, Daniel Hritzkiv submitted commit 9ec117d0bcd97b6e0fc21083c9d995ebc4309527 which modified LazyDict and changed the MergeMeshes code to remove meshes referred by other entities. If I revert that commit, exporting to gltf2 works again (although with the problem that commit was supposed to fix).

My suspicion is that the use of LazyDict::Remove is being too aggressive and corrupting memory. I haven't yet dug into it yet to submit a patch but research is ongoing.

**Comment by dhritzkiv** [comment]
Hey @Misterblue. Sorry about that. It could be that a mesh is referenced in multiple nodes, and when merging that mesh into the main mesh's primitives, its reference is lost for other nodes. There aren't any checks for that, currently, as I didn't think it was possible to have any top-level meshes be referenced more than once.

Do you have an example model that I can test with?

**Comment by Misterblue** [comment]
@dhritzkiv, yes. I think the problem is indeed that multiple meshes refer to the same primitive mesh. Rather than deleting the scene mesh as soon as the primitive is created, there maybe should be code to make a pass looking for unused meshes after the meshes have been merged.

An example scene that fails the conversion to gltf2 is at http://files.misterblue.com/oars/Atropia/Atropia_12.dae .

**Comment by kimkulling** [comment]
Thanks for the report. I guess this behaviour is some kind of instancing. At the moment assimp does not deal fine with these kind of mesh-instancing as well.

**Comment by wlinna** [comment]
I get  `invalid vector<T> subscript` error on [this line](https://github.com/assimp/assimp/blob/master/code/glTF2Exporter.cpp#L844):

    Ref<Mesh> mesh = node->meshes.at(m);

I sprinkled print calls all over, and discovered that two calls to `node->meshes.erase(node->meshes.begin() + mm);` occurred. The next time the call `node->meshes.at(m)` happened, `m` was out of bounds:

`m = 2, nMeshes = 4, size() = 2`

Could it be that all it takes to fix this error is to update `m` or to check `node->meshes.size()` in the for-loop? I haven't had the time to understand the code yet, so I don't know.

**Comment by wlinna** [comment]
Hmm. I actually could not reproduce this with the latest code (I was using my own version with some modifications)

**Comment by wlinna** [comment]
My initial hope that a simple bounds check would suffice turned out to be wrong. What I ended up doing was that before `MergeMeshes` I move all instanced meshes to their own nodes, and copy the necessary fields. Then there's no need to modify `MergeMeshes`.

My change doesn't work verbatim with the upstream Assimp, because I've made other changes to Assimp to suit our needs, and my solution takes advantage of those changes. The code should, however, be reasonably easy to adapt to work upstream. I'll consider rewriting this for upstream and making a PR, but here's the backbone of the solution before that happens.

```
void glTF2Exporter::SplitInstanceNodes() {
    auto initialNodeCount = mAsset->nodes.Size();
    for (unsigned int ni = 0; ni < initialNodeCount; ++ni) {
        Ref<Node> node = mAsset->nodes.Get(ni);
        unsigned int nMeshes = static_cast<unsigned int>(node->meshes.size());

        if (nMeshes <= 1) {
            continue;
        }

        for (auto mi = 0; mi < node->meshes.size();) {
            auto meshRef = node->meshes[mi];
            auto handle = node->meshHandles[mi];
            auto instanceCount = mInstanceInfo[meshRef->oIndex].size();
            if (instanceCount > 1) {
                auto name = mAsset->FindUniqueID(handle->mName.C_Str(), "node");
                auto newNode = mAsset->nodes.Create(name);
                node->parent->children.push_back(newNode);
                newNode->matrix = node->matrix;
                newNode->rotation = node->rotation;
                newNode->scale = node->scale;
                newNode->translation = node->translation;
                newNode->meshes.push_back(meshRef);
                node->meshes.erase(node->meshes.begin() + mi);
                continue;
            }
            else {
                ++mi;
            }
        }
    }
}
```
I haven't tested this much yet though, but at least the problematic model converts to glTF now.

**Comment by tellypresence** [comment]
Please add "Export" label to this issue

**Comment by tellypresence** [comment]
@Misterblue Above 3D model link http://files.misterblue.com/oars/Atropia/Atropia_12.dae no longer valid, please provide updated link

**Comment by tellypresence** [comment]
@Misterblue Is this still a problem?

**Comment by Misterblue** [comment]
Don't know what happened to the .dae file. I'll have to check my backups.

I haven't used assimp in a long time and I don't have a way of testing if this is still a problem. If the code hasn't changed, then it probably is.

**Comment by tellypresence** [comment]
Closing old issue for now, please reopen if this is still a problem
